Geo: Add validation of shapes to ShapeBuilders

cbuescher opened the pull request:
Currently the validation of geo shapes is taking place in the various parse() methods in ShapeBuilder. With the search refactoring, we can no longer rely on shapes being parsed from json, so the shapes also need to be validated when just using the java api.

A lot of validation concerns the number of points a shape needs to have in order to be valid (e.g. LineString >= 2 points). This is not possible with current the builders where points can be added one by one, so some no-arg constructors are changed to require mandatory parameters and to validate those at construction time. To help with constructing longer lists of points, a new utility PointsListBuilder is instroduces which can produce list of coordinates accepted by most of the other shape builder constructors. 

Also adding tests for invalid shape exceptions to the already existing shape builder tests.

Relates to #14416

cbuescher committed:
Geo: Add validation of shapes to shape builders

So far the validation of geo shapes was only taking place in the
parse methods in ShapeBuilder. With the recent refactoring we no
longer can rely on shapes being parsed from json, so the same kind
of validation should take place when just using the java api.

A lot of validation concerns the number of points a shape needs to
have in order to be valid. Since this is not possible with current
builders where points can be added one by one, the builder constructors
are changed to require the mandatory parameters and validate those
already at construction time. To help with constructing longer lists
of points, a new utility PointsListBuilder is instroduces which can
produce list of coordinates accepted by most of the other shape builder
constructors.

Also adding tests for invalid shape exceptions to the already existing
shape builder tests.

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/PointListBuilder.java:
```diff
@@ -0,0 +1,98 @@
+ * Enables chaining of individual points either as long/lat pairs
+ * or as {@link Coordinate} elements, arrays or collections.
+ */
+public class PointListBuilder {
```
Super nit: Can we refactor this class name to CoordinateListBuilder? That way the naming is a bit more consistent, such as:
```
CoordinateListBuilder cl = new CoordinateListBuilder();
int size = in.readVInt();
for (int i=0; i < size; i++) {
  cl.coordinate(readCoordinateFrom(in));
}
```

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/PointListBuilder.java:
```diff
@@ -0,0 +1,98 @@
+     * @param latitude latitude of the coordinate
+     * @return this
+     */
+    public PointListBuilder point(double longitude, double latitude) {
```
same as above. refactor ```point``` to ```coordinate```

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/PointListBuilder.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * @return the current list of points
+     */
+    public List<Coordinate> list() {
```
then maybe we can rename this to ```build```? And call build inside the ShapeBuilder constructor?
Something like:
```
public static final PolygonBuilder PROTOTYPE = new PolygonBuilder(new CoordinatesBuilder().coordinates(0.0, 0.0).coordinates(0.0, 1.0).coordinates(1.0, 0.0).coordinates(0.0, 0.0));
```

nknize commented on core/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java:
```diff
@@ -64,55 +65,40 @@
 
-    public PolygonBuilder() {
-        this(Orientation.RIGHT);
+    public PolygonBuilder(List<Coordinate> points, Orientation orientation) {
```
Take in a ```CoordinatesBuilder```? and call ```build()``` inside the constructor?

nknize commented:
Logic LGTM. Left a few minor comments.

cbuescher committed:
Rename PointCollectionBuilder to CoordinatesBuilder

Also renaming internal methods to reflect that they are dealing with
jts coordinates. Also renamed the list() to build() method for creating
the coordinates lists and adding constructors to PolygonBuilder that
take CoordinatesBuilders and implicitely call build() on them.

cbuescher commented:
Thanks for the review @nknize, I addressed your comments by renaming the new helper for chaining points to CoordinatesBuilder (also changing the method names). I also changed the name of the common supertype of LineStringBuilder and MultiPointBuilder from PointCollection to CoordinateCollection to reflect that it wraps a List<Coordinate>. I left the MultiPointBuilder name as is since that is the name of the GeoShape it creates.

I have a minor question left concerning the "coerce" option for PolygonBuilder. So far polygon shells and holes could be coerced from un-closed coordinate lists in ShapeBuilder#parseLinearRing(). I'm not sure if we need this option for the PolygonBuilder as well, that's why I added an aditional constructor and a hole() method with the "coerce" option, but both are basically unused in other than in new tests that I introduced, so I'm wondering if those options are needed when using the Java API or if we rather force users to close shells/holes themselves explicitely. What do you think?

nknize commented:
> I'm wondering if those options are needed when using the Java API

Without them the Java API would be inconsistent with the REST parser, correct? If so then I think we should keep them for consistency?

cbuescher commented:
As far as I can see the REST parsers don't really have a dedicated "coerce" parameter, that flag is retrieved from the FieldMapper if one is present, and ShapeBuilder#parseLinearRing() should add the missing point before creating the PolygonBuilder, so I'm not sure if that option is needed in the PolygonBuilder.

nknize commented:
Ah yes, I was thinking of the QueryBuilders. You're right, the mapper should contain the coerce option. I don't think its necessary to carry it in the PolygonBuilder, but that can be done in a separate PR? I'm good w/ this as is.

cbuescher commented:
Okay, thanks, will merge then.
